pkg/logql/log/parser.go

```diff
@@ -156,32 +165,68 @@ func (j *JSONParser) parseLabelValue(key, value []byte, dataType jsonparser.Valu
 
 	// snapshot the current prefix position
 	prefixLen := len(j.prefixBuffer)
-	j.prefixBuffer = append(j.prefixBuffer, byte(jsonSpacer))
-	j.prefixBuffer = appendSanitized(j.prefixBuffer, key)
-	keyString, ok := j.keys.Get(j.prefixBuffer, func() (string, bool) {
-		if j.lbs.BaseHas(string(j.prefixBuffer)) {
-			j.prefixBuffer = append(j.prefixBuffer, duplicateSuffix...)
+	j.prefixBuffer = append(j.prefixBuffer, key)
+
+	sanitized := j.buildSanitizedPrefixFromBuffer()
+	keyString, ok := j.keys.Get(sanitized, func() (string, bool) {
+		if j.lbs.BaseHas(string(sanitized)) {
+			j.prefixBuffer[prefixLen] = append(key, duplicateSuffix...)
 		}
-		if !j.parserHints.ShouldExtract(string(j.prefixBuffer)) {
+
+		keyPrefix := j.buildSanitizedPrefixFromBuffer()
+		if !j.parserHints.ShouldExtract(string(keyPrefix)) {
 			return "", false
 		}
 
-		return string(j.prefixBuffer), true
+		return string(keyPrefix), true
 	})
 
+	if j.captureJSONPath {
+		jsonPath := j.buildJSONPathFromPrefixBuffer()
+		j.lbs.SetJSONPath(keyString, jsonPath)
+	}
+
 	// reset the prefix position
 	j.prefixBuffer = j.prefixBuffer[:prefixLen]
 	if !ok {
 		return nil
 	}
 
 	j.lbs.Set(ParsedLabel, keyString, readValue(value, dataType))
+
 	if !j.parserHints.ShouldContinueParsingLine(keyString, j.lbs) {
 		return errLabelDoesNotMatch
 	}
 	return nil
 }
 
+func (j *JSONParser) buildSanitizedPrefixFromBuffer() []byte {
+	j.santizedPrefixBuffer = j.santizedPrefixBuffer[:0]
+
+	for i, part := range j.prefixBuffer {
+		if len(bytes.TrimSpace(part)) == 0 {
+			continue
+		}
+
+		if i > 0 && len(j.santizedPrefixBuffer) > 0 {
+			j.santizedPrefixBuffer = append(j.santizedPrefixBuffer, byte(jsonSpacer))
+		}
+		j.santizedPrefixBuffer = appendSanitized(j.santizedPrefixBuffer, part)
+	}
+
+	return j.santizedPrefixBuffer
+}
+
+func (j *JSONParser) buildJSONPathFromPrefixBuffer() []string {
+	jsonPath := make([]string, 0, len(j.prefixBuffer))
+	for _, part := range j.prefixBuffer {
+		partStr := unsafe.String(unsafe.SliceData(part), len(part)) // #nosec G103 -- we know the string is not mutated
```

Comment:
the parts of the path will not change, so I've optimized this to use `unsafe.String`, which saves ~20% in terms of allocations.